internal/pkg/agent/application/actions/handlers/handler_action_policy_change.go

```diff
@@ -118,104 +118,160 @@ func (h *PolicyChangeHandler) Watch() <-chan coordinator.ConfigChange {
 	_, _ = io.Copy(io.Discard, resp.Body)
 	resp.Body.Close()
 
-	reader, err := fleetToReader(h.agentInfo, h.config)
+	return nil
+}
+
+// updateFleetConfig copies the relevant Fleet client settings from src on dst. The destination struct is modified in-place
+func updateFleetConfig(log *logger.Logger, src remote.Config, dst *remote.Config) {
+	dst.Protocol = src.Protocol
+	dst.Path = src.Path
+	dst.Host = src.Host
+	dst.Hosts = src.Hosts
+
+	// Empty proxies from fleet are ignored. That way a proxy set by --proxy-url
+	// it won't be overridden by an absent or empty proxy from fleet-server.
+	// However, if there is a proxy sent by fleet-server, it'll take precedence.
+	// Therefore, it's not possible to remove a proxy once it's set.
+
+	if src.Transport.Proxy.URL == nil ||
+		src.Transport.Proxy.URL.String() == "" {
+		log.Debug("proxy from fleet is empty or null, the proxy will not be changed")
+	} else {
+		// copy the proxy struct
+		dst.Transport.Proxy = src.Transport.Proxy
+
+		// replace in dst the attributes that are passed by reference within the proxy struct
+
+		// Headers map
+		dst.Transport.Proxy.Headers = map[string]string{}
+		for k, v := range src.Transport.Proxy.Headers {
+			dst.Transport.Proxy.Headers[k] = v
+		}
+
+		// Proxy URL
+		urlCopy := *src.Transport.Proxy.URL
+		dst.Transport.Proxy.URL = &urlCopy
+
+		log.Debug("received proxy from fleet, applying it")
+	}
+}
+
+func (h *PolicyChangeHandler) handlePolicyChange(ctx context.Context, c *config.Config) (err error) {
+	cfg, err := configuration.NewFromConfig(c)
 	if err != nil {
-		return errors.New(
-			err, "fail to persist new Fleet Server API client hosts",
-			errors.TypeUnexpected, errors.M("hosts", h.config.Fleet.Client.Hosts))
+		return errors.New(err, "could not parse the configuration from the policy", errors.TypeConfig)
+	}
+
+	// validate Fleet connectivity with the new configuration
+	var validatedConfig *remote.Config
+	validatedConfig, err = h.validateFleetServerHosts(ctx, cfg)
+	if err != nil {
+		return fmt.Errorf("error validating Fleet client config: %w", err)
+	}
+
+	// persist configuration
+	err = saveConfig(h.agentInfo, cfg, h.store)
+	if err != nil {
+		return fmt.Errorf("saving FleetClientConfig: %w", err)
+	}
+
+	// apply the new configuration to the current clients
+	err = h.applyFleetClientConfig(validatedConfig)
+	if err != nil {
+		return fmt.Errorf("applying FleetClientConfig: %w", err)
+	}
+
+	return nil
+}
+
+func (h *PolicyChangeHandler) applyFleetClientConfig(validatedConfig *remote.Config) error {
+	if validatedConfig == nil || len(h.setters) == 0 {
+		// nothing to do for fleet hosts
+		return nil
+	}
+
+	h.config.Fleet.Client = *validatedConfig
+
+	// the config has already been validated, no need for error handling
+	fleetClient, err := client.NewAuthWithConfig(
+		h.log, h.config.Fleet.AccessAPIKey, *validatedConfig)
+	if err != nil {
+		return fmt.Errorf("creating new fleet client with updated config: %w", err)
+	}
```

Comment:
[Suggestion]
The error is being handles, so the comment should reflect it.

```suggestion
	// the config has already been validated, it should not error.
	fleetClient, err := client.NewAuthWithConfig(
		h.log, h.config.Fleet.AccessAPIKey, *validatedConfig)
	if err != nil {
		return fmt.Errorf("creating new fleet client with updated config: %w", err)
	}
```